draft-ietf-quic-transport.md

```diff
@@ -1542,6 +1542,80 @@ lifetime of a connection, especially in response to connection migration
+handshake are also authenticated by the cryptographic handshake.
+
+Each endpoint includes the value of the Source Connection ID field from the
+first Initial packet it sent in the initial_source_connection_id transport
+parameter; see {{transport-parameter-definitions}}. A server includes the
+Destination Connection ID field from the first Initial packet it received from
+the client in the original_destination_connection_id transport parameter; if
+the server sent a Retry packet this refers to the first Initial packet received
+before sending the Retry packet. If it sends a Retry packet, a server also
+includes the Source Connection ID field from the Retry packet in the
+retry_source_connection_id transport parameter.
+
+The values provided by a peer for these transport parameters MUST match the
+values that an endpoint used in the Destination and Source Connection ID fields
+of Initial packets that it sent. Including connection ID values in transport
+parameters and verifying them ensures that that an attacker cannot influence
+the choice of connection ID for a successful connection by injecting packets
+carrying attacker-chosen connection IDs during the handshake. An endpoint MUST
+treat any of the following as a connection error of type PROTOCOL_VIOLATION:
+
+* absence of the initial_source_connection_id transport parameter from either
+  endpoint,
+
+* absence of the original_destination_connection_id transport parameter from
+  the server,
+
+* absence of the retry_source_connection_id transport parameter from the server
+  after receiving a Retry packet,
+
+* presence of the retry_source_connection_id transport parameter when no Retry
+  packet was received, or
+
+* a mismatch between values received from a peer in these transport parameters
+  and the value sent in the corresponding Destination Connection ID fields of
+  Initial packets.
+
+If a zero-length connection ID is selected, the corresponding transport
+parameter is included with a zero-length value.
+
+{{fig-auth-cid}} shows the connection IDs that are used in a complete
+handshake. The exchange of Initial and Retry packets is shown, plus the later
+exchange of 1-RTT packets that includes the connection ID established during
+the handshake.
+
+~~~
+Client                                                  Server
+
+Initial: DCID=S1, SCID=C1 ->
+                                   <- Retry*: DCID=C1, SCID=S2
+Initial*: DCID=S2, SCID=C1 ->
+                                  <- Initial: DCID=C1, SCID=S3
+                             ...
+1-RTT: DCID=S3 ->
+                                             <- 1-RTT: DCID=C1
+~~~
+{: #fig-auth-cid title="Use of Connection IDs in a Handshake"}
+
+For the handshake in {{fig-auth-cid}} the client sets the value of the
+initial_source_connection_id transport parameter to `C1`. If the server sends a
+Retry packet (that is, the packets marked with a '*' are sent), it sets
```

Comment:
I really like the diagram, but I would prefer to split it in two. There are two possible handshakes described here: without and with retry involved. I think that the `the packets marked with a '*' are sent` caveat is confusing, because when those packets aren't sent, the `1-RTT: DCID=S3` needs to be replaced by `1-RTT: DCID=S2`. It might be simpler to have two diagrams and two examples.